pkg/driver/directory_provisioner.go

```diff
@@ -0,0 +1,144 @@
+package driver
+
+import (
+	"context"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog"
+
+	"github.com/kubernetes-sigs/aws-efs-csi-driver/pkg/cloud"
+)
+
+type DirectoryProvisioner struct {
+	mounter              Mounter
+	cloud                cloud.Cloud
+	osClient             OsClient
+	deleteProvisionedDir bool
+}
+
+func (d DirectoryProvisioner) Provision(ctx context.Context, req *csi.CreateVolumeRequest, uid, gid int) (*csi.Volume, error) {
+	var provisionedPath string
+
+	var fileSystemId string
+	volumeParams := req.GetParameters()
+	if value, ok := volumeParams[FsId]; ok {
+		if strings.TrimSpace(value) == "" {
+			return nil, status.Errorf(codes.InvalidArgument, "Parameter %v cannot be empty", FsId)
+		}
+		fileSystemId = value
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", FsId)
+	}
+	klog.V(5).Infof("Provisioning directory on FileSystem %s...", fileSystemId)
+
+	localCloud, roleArn, err := getCloud(d.cloud, req.GetSecrets())
+	if err != nil {
+		return nil, err
+	}
+
+	mountOptions, err := getMountOptions(ctx, localCloud, fileSystemId, roleArn)
+	if err != nil {
+		return nil, err
+	}
+	target := TempMountPathPrefix + "/" + uuid.New().String()
+	if err := d.mounter.MakeDir(target); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+	}
+	if err := d.mounter.Mount(fileSystemId, target, "efs", mountOptions); err == nil {
+		// Extract the basePath
+		var basePath string
+		if value, ok := volumeParams[BasePath]; ok {
+			basePath = value
+		}
+
+		rootDirName := req.Name
+		provisionedPath = basePath + "/" + rootDirName
+
+		klog.V(5).Infof("Provisioning directory at path %s", provisionedPath)
+
+		// Grab the required permissions
+		perms := os.FileMode(0755)
+		if value, ok := volumeParams[DirectoryPerms]; ok {
+			parsedPerms, err := strconv.ParseUint(value, 8, 32)
+			if err == nil {
+				perms = os.FileMode(parsedPerms)
+			}
+		}
+
+		klog.V(5).Infof("Provisioning directory with permissions %s", perms)
+
+		provisionedDirectory := path.Join(target, provisionedPath)
+		err := d.osClient.MkDirAllWithPerms(provisionedDirectory, perms, uid, gid)
```

Comment:
How does a pod that uses such a PV get access to the directory? In other words, how user or Kubernetes know that it should run it with allocated UID/GID? For AccessPoints, EFS itself makes sure that any UID/GID is translated to the allocated one.

IMO it would make more sense not to allocate any UID/GID at all and expect `DirectoryPerms: 777` in the storageClass or `fsGroup` in Pods.

Reply:
Yes I agree, I've made the change to make the permissions 777 by default but the options is still there in StorageClass to override them if people want to for whatever reason (I'm sure there are use-cases where it makes sense)

Reply:
The documentation is now kind of confusing now. I don't know how to make it better, maybe in a subsequent PR. It should say that for `efs-dir` it is recommended to have `DirectoryPerms: 777` and omit `gidRangeStart` / `gidRangeEnd`, so a pod can access its volume and on the other hand, `efs-ap` should have `DirectoryPerms: 770` and some `gidRangeStart`/`gidRangeEnd`.

All options are available in both modes if users really know what they are doing.

Reply:
Have made a change to the README to add our recommendations, let me know what you think